File: projects/temp/cards/bindings/c_sharp/class_1/kabuki_card_games/Deck.cs
/** kabuki::cards
    @file       ~/kabuki::cards Library/Deck.cs
    @author  Cale McCollough <cale.mccollough@gmail.com>
    @license Copyright (C) 2017 Cale McCollough <calemccollough.github.io>;
             All right reserved (R). Licensed under the Apache License, Version 
             2.0 (the "License"); you may not use this file except in 
             compliance with the License. You may obtain a copy of the License 
             [here](http://www.apache.org/licenses/LICENSE-2.0). Unless 
             required by applicable law or agreed to in writing, software
             distributed under the License is distributed on an "AS IS" BASIS,
             WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 
             implied. See the License for the specific language governing 
             permissions and limitations under the License.
*/

namespace DojoCardGames
{
    using System;
    using System.Collections.Generic;
    using System.Diagnostics;

    /** A deck of playing cards.
        Termenology:
        - Pack: A pack of cards has 52 cards without jokers and 54 with jokers.
        - Deck: A deck can contain multiple packs of cards.
        - Stock: The main stack of cards is called the stock of cards.

        @see Card class
    */
    public class Deck
    {
        //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

        public const int PackSize = 52,         //< The number of cards in a single pack of cards excluding jokers.
            PackSizeWithJokers = 54;            //< Flag for if the deck doesn't have jokers.

        public const bool Jokers = true,        //< Deck contains jokers.
            NoJokers = false;                   //< Deck contains no jokers.

        /** Constructor. */
        public Deck (bool newDeckHasJokers = false, int newSuitType = Card.French, int newNumPacks = 1, 
            int[] newFaceValues = null, int[] newSuitValues = null)
        {
            hasJokers = newDeckHasJokers;
            numPacks = newNumPacks < 1 ? 1 : newNumPacks;

            int numCardsInPack = hasJokers ? PackSizeWithJokers : PackSize;

            deckSize = numCardsInPack * numPacks;

            // Create the cards.
            cards = new List<Card> ();

            if (newSuitType < 1)
                suitType = 1;
            else if (newSuitType > Card.NumSuitTypes)
                suitType = Card.NumSuitTypes;
            else
                suitType = newSuitType;

            faceValues = new int[Card.NumPips];
            suitValues = new int[Card.NumSuits];

            if (newFaceValues == null || newFaceValues.Length != Card.NumPips)
            {
                for (int i = 0; i < Card.NumPips; i++)
                    faceValues[i] = i;
            }
            else
                for (int i = 0; i < Card.NumPips; i++)
                    faceValues[i] = newFaceValues[i];

            if (newSuitValues == null || newSuitValues.Length != Card.NumSuits)
            {
                for (int i = 1; i <= Card.NumSuits; i++)
                    suitValues[i - 1] = i;
            }
            else
                for (int i = 0; i < Card.NumSuits; i++)
                    suitValues[i] = newSuitValues[i];

            String faceValuesString = "",
                suitValuesString = "";

            for (int i = 0; i < Card.NumPips; i++)
                faceValuesString += faceValues[i] + ", ";

            for (int i = 0; i < Card.NumSuits; i++)
                suitValuesString += suitValues[i] + ", ";

            for (int i = 1; i <= numPacks; i++)
                AddPackToDeck ();
        }

        /** Function the returns if the stock is empty. */
        public bool IsEmpty
        {
            get { return cards.Count == 0; }
        }

        /** Shuffles the stock. */
        public void Shuffle ()
        {
            int numberOfCards = NumCards;

            var shuffledCards = new List<Card>(NumCards);

            for (int i = 0; i < numberOfCards; i++)

                shuffledCards.Add (DrawRandomCard ());

            cards = shuffledCards;
        }

        /** Returns the number of cards in the cards. */
        public int NumCards
        {
            get { return cards.Count; }
        }

        /** Returns true if this deck has jokers in it. */
        public bool HasJokers
        {
            get { return hasJokers; }
        }

        /** Returns the Card at the specified index without removing it.
            @return Returns a black joker if the index is out of bounds.
        */
        public Card GetCard (int index)
        {
            if (index < 0 || index >= cards.Count)
                return null;

            return cards[index];
        }

        /** Draws a random card from the cards. */
        public Card DrawCard ()
        {
            if (cards.Count == 0) // Then there are no cards in the cards.
                return null;

            int lastIndex = cards.Count - 1;
            /*< Golden rule: If you use a variable in a function more than once, store a local copy to get a performance 
                increase. */

            var nextCard = cards[lastIndex];     //< Store the top card on the cards in a temp variable.
            cards.RemoveAt (lastIndex);          //< Now remove the card from the data structure (i.e. the List)

            return nextCard;                    //< The you return the temp variable... You'll see this pattern a lot.
        }

        /** Returns a Card witht he given pip and suit values.
            @return Return a Red Joker with suitValue of 3 (diamons) if the cards doesn't contain the card. */
        public Card DrawCard (int pipValue, int suitValue)
        {
            // We have to search the whole cards with a for loop.
            for (int i = 0; i < cards.Count; i++)
            {
                var currentCard = cards[i];

                if (currentCard.Equals (pipValue, suitValue))
                {
                    cards.RemoveAt (i);
                    return currentCard;
                }
            }
            return null;
        }

        /** Draws a random card from the cards. */
        public Card DrawRandomCard ()
        {
            if (cards.Count == 0) //< Then there are no cards in the cards.
                return null;

            int randomIndex = randomGenerator.Next (NumCards);
            var returnCard = cards[randomIndex];    //< Store the random card as a temp variable.
            cards.RemoveAt (randomIndex);           //< Remove the randomIndex we saved earlier.
            return returnCard;                      //< Then return the temp variable.
        }

        /** Gets the number of packs in the deck. */
        public int NumPacks
        {
            get  {  return numPacks; }
        }

        /** Returns the maximum number of cards in a new full deck.
            @warning    The number of cards in the deck may exceed this number. */
        public int DeckSize
        {
            get { return deckSize; }
        }

        /** Adds a packs's worth of cards with this deck. */
        public void AddPackToDeck ()
        {
            /** The order of the pips are A,2,3,4,5,6,7,8,9,10,J,Q,K, and don't want any Jokers. We need to 
                start at 1, and loop through 13 times. */
				
            for (int suit = Card.Ace; suit <= 4; suit++)                    //< 
            {
                for (int pipValue = 1; pipValue <= Card.King; pipValue++) 	// Once for each of the 13 pip values 
                {
                    /** This is called a nested for loop. It makes it so that our cards will be sorted by suit and 
                        value. Note: If we had switched the order of the nested for loops, it would sort the cards by 
                        value then suit.
                    */

                    var newCard = new Card (pipValue, suit, faceValues[pipValue], suitValues[suit - 1], suitType);
                    cards.Add (newCard);
                }
            }

            if (hasJokers)
            {
                cards.Add (new Card (Card.Joker, Card.Red));
                cards.Add (new Card (Card.Joker, Card.Black));
            }
        }

        /** Returns a String representation of this Object. */
        public override String ToString ()
        {
            String returnString = "Deck: numCards: ";

            returnString += cards.Count + "\n";

            for (int i = 0; i < cards.Count; ++i)
            {
                returnString += cards[i].ToString ();

                if (i != cards.Count)
                    returnString += ", ";
                else
                    returnString += "\n";
            }

            return returnString;
        }
        //--------------------------------------------------------------------------------------------------------------

        private int numPacks,                   //< The number of decks.
            deckSize,                       	//< The size of a full deck.
            suitType;                           //< The suit type for this Deck
        private int[] suitValues,               //< The map of suit values.
            faceValues;                         //< The array of faceValues.
        private bool hasJokers;                 //< Stores if this deck has jokers or not.

        private List<Card> cards;               /*< The Array of Card objects.
													The <> is a C++ thing. Its called a template. We just made a Array of Card. */
		/** A Random number generator. */
        private Random randomGenerator = new Random ();
    }
















    public class DeckTests
    {
        public static void RunTests ()
        {
            Debug.Write ("\nTesting Deck Class\n" +
                "*******************************************************************************\n\n" + 
                "Creating test Deck...\n\n");
            
            Deck stock = new Deck ();

            Debug.Assert (stock.NumCards == 52, "Error: There where not 52 stock!\n" + stock.ToString());

            Debug.Write (stock.ToString () + "\n\nShuffling Deck...\n");
            
            stock.Shuffle ();

            Debug.WriteLine ("\n\n" + stock.ToString () + 
                "\n\nTesting DrawCard () by drawing 42 stock...\n");

            for (int i = 0; i < 42; i++)
                Debug.Write (stock.DrawCard ().ToString () + ", ");

            Debug.WriteLine ("\n\nPrinting last 10 stock...\n\n" + stock.ToString () + 
                "\n\nTesting DrawRandomCard () by drawing 5 stock...\n");

            for (int i = 0; i < 5; i++)
                Debug.Write (stock.DrawRandomCard ().ToString () + ", ");

            int numLoops = 10,
                numExpectedNullstock = numLoops - stock.NumCards,
                counter = 0;

            Debug.WriteLine ("\n\nPrinting last " + numExpectedNullstock + " stock...\n\n" + stock.ToString () +
                "\n\nDrawing 10 more random stock and expecting 5 null stock...\n");

            String drawnstock = "";
            
            for (int i = 1; i <= numLoops; i++)
            {
                var tempCard = stock.DrawRandomCard ();

                if (tempCard != null)
                {
                    Debug.WriteLine (i + ", ");
                    drawnstock += tempCard.ToString () + ", ";
                }
                else
                    counter++;
            }
            Debug.Assert (counter == numExpectedNullstock, "Error: There were supposed to be " + numExpectedNullstock + " null stock and got " + counter + "!\nDraw stock:\n" + drawnstock + "\n" + stock.ToString ());
            
            Debug.Assert (stock.IsEmpty, "Error: After trying to take more stock than the stock had, the stock was not empty!");

            Debug.Write ("\n\nPrinting empty Deck...\n\n" + stock.ToString () + 
                "\n\nShuffling and printing stock...\n");

            stock.Shuffle();

            Debug.WriteLine (stock.ToString () + "\n\n");

            Debug.WriteLine ("Done testing the Deck class.\n");
        }
    }
}

File: projects/igeek/source/cards/duplicate/deck.cc
/** kabuki::cards
    @file       ~/source/kabuki/cards/Deck.cc
    @author  Cale McCollough <cale.mccollough@gmail.com>
    @license Copyright (C) 2017 Cale McCollough <calemccollough.github.io>;
             All right reserved (R). Licensed under the Apache License, Version 
             2.0 (the "License"); you may not use this file except in 
             compliance with the License. You may obtain a copy of the License 
             [here](http://www.apache.org/licenses/LICENSE-2.0). Unless 
             required by applicable law or agreed to in writing, software
             distributed under the License is distributed on an "AS IS" BASIS,
             WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 
             implied. See the License for the specific language governing 
             permissions and limitations under the License.
*/

#include "Deck.h"
#include "CardStack.h"

using namespace KabukiCardGames;

const int Deck::defaultSuitValues[4] = { 1, 2, 3, 4 };

const String Deck::defaultDeckArtDirectory = "../../../Images/FrontDeckArt/vectorized-playing-cards/";
const String Deck::defaultRearDeckImage = "../../../Images/RearDeckArt/CoderDojoDeckRearImage.svg";

const Deck Deck::standard = Deck ();
const Deck Deck::standardWithJokers = Deck (false);

Deck::Deck (bool deckContainsJokers, int acesAreHigh, Card::SuitCulture thisSuitCulture, String& rearDeckImageFilename, const String& deckArtDirectory) :
    hasJokers (deckContainsJokers),
    acesHigh (acesAreHigh),
    suitCulture (thisSuitCulture),
    numCards (deckContainsJokers ? defaultNumCardsWithJokers : defaultNumCards),
    lowestCardValue (acesAreHigh == 0 ? 1 : 2),
    highestCardValue (acesAreHigh == 0 ? 13 : 14)
{
    initialise (rearDeckImageFilename, deckArtDirectory);
}

void Deck::initialise (String& rearDeckImageFilename, const String& deckArtDirectory)
{
    if (deck.size () == 0)
        deck.clear ();
    // Depending on if aces are high, the value of an Ace will be either 1 followed by the 2 card, or 14.
    int startValue = acesHigh ? 1 : 2;
    const String& suitCultureStringAddress = Card::suitCultureStrings[suitCulture];

    // First we want to start by creating the Aces because their pipValue is 1, but might have a faceValue of 14.
    int aceValue = acesHigh ? 14 : 1;

    for (int suit = 1; suit <= 4; ++suit)
        deck.add (new Card (1, 1, aceValue, suitValues[suit], (Card::Suit) suit, Card::suitCultureStrings[suit], deckArtDirectory);

    // This is a nested for loop. It makes it so that our deck will be sorted by suit and value.
    for (int suit = 1; suit <= 4; ++suit) // If we had switched the order of the for loops, it would sort it by value then suit.
        for (int cardValue = 2; cardValue <= 13; ++cardValue) // There are 13 different face values { 2,3,4,5,6,7,8,9,10,J,Q,K,A }
            deck.add (new Card (cardValue, (Card::Suit)suit, this)); // The (Card::Suit) "casts" the suitValue to a Card::Suit.

    rearImage = ImageCache::getFromFile (rearDeckImageFilename);

    if (rearImage.isNull ())    //< null means that there was an error so try to load the default deck.
        ImageCache::getFromFile (defaultRearDeckImage);
}



Deck& Deck::operator= (const Deck& other)
{
}

int Deck::setSuitValues (int value1, int value2, int value3, int value4)
{
    if (value1 < 1)
        return -1;
    if (value1 > 4)
        return 1;

    if (value2 < 1)
        return -2;
    if (value2 > 4)
        return 2;

    if (value3 < 1)
        return -3;
    if (value3 > 4)
        return 3;

    if (value4 < 1)
        return -4;
    if (value4 > 4)
        return 4;

    suitValues[1] = value1;
    suitValues[2] = value2;
    suitValues[3] = value3;
    suitValues[4] = value4;
}

int Deck::getSuitValue (int suit)
{
    if (suit < 1 || suit > 4)
        return -1;
    return suitValues[suit];
}

bool Deck::deckHasJokers ()
{
    return hasJokers;
}

int Deck::getNumCards ()
{
    return numCards;
}

Card* Deck::getCard (int index)
{
    if (index < 0 || index >= deck.size ())
        return nullptr;
    return deck[index];
}

int Deck::setRearImage (const File& thisFile)
{
    Image tempImage = ImageCache::getFromFile (thisFile);

    if (tempImage.isNull)
        return 0;

    rearImage = tempImage;

    return 1;
}

int Deck::checkDeckArtFolder (String& deckArtFolder)
{
    File artFolder = File (deckArtFolder); //< Make a file fromt he deckArtFolder string.

    // Before we go trying to change all of the images, lets first check to see if the deckArtFolder is a valid folder.
    if (!artFolder.isDirectory ())
        return 1;

    // Next, we need to check to see if the folder contains 54 .svg images.
    Array<File> cardImageFiles;
    if (artFolder.findChildFiles (cardImageFiles, File::TypesOfFileToFind::findFiles, false, ".svg") != 54)
        return 54;
    // Note: I'm assuming that the deckArtFolder is sorted by name.
    // Then check to see if all of the images were nammed correctly.
    for (int suit = 1; suit <= 4; ++suit)
    {
        for (int pipValue = 1; pipValue <= 13; ++pipValue)
        {
            int index = suit*pipValue;

            String expectedFileName = String (pipValue) + "-" + String (suit) + ".svg"; // Examples "1-2.svg" = Ace of Diamons "13-4" = King of Spades

            if (cardImageFiles[index].getFileName () != expectedFileName)
                return -1 * index;
        }
    }
    if (cardImageFiles[52].getFileName () != "0-1.svg")
        return -53;
    if (cardImageFiles[53].getFileName () != "0-2.svg")
        return -54;
}

int Deck::setDeckArt (String& deckArtFolder)
{
    int returnValue = Deck::checkDeckArtFolder ();

    // Now we know that all of the files are in the directory and are nammed correctly. Now lets actually change the images.
    for (int suit = 1; suit <= 4; ++suit)
    {
        for (int pipValue = 1; pipValue <= 13; ++pipValue)
        {
            int index = suit*pipValue;
            String filename = String (pipValue) + "-" + String (suit) + ".svg";

            int returnValue = deck[index]->setCardImage (ImageCache::getFromFile (filename));

            if (returnValue != 0)
                return returnValue;
        }
    }
    return 0;
}

String Deck::getSuitString (Card::Suit suit)
{
    switch (suitCulture)
    {
        case Card::french:
        {
            switch (suit)
            {
            case 1: return "Clubs";
            case 2: return "Diamonds";
            case 3: return "Hearts";
            case 4: return "Spades";
            }
            break;
        }
        case Card::german:
        {
            switch (suit)
            {
            case 1: return "Acorns";
            case 2: return "Bells";
            case 3: return "Hearts";
            case 4: return "Lieves";
            }
            break;
        }
        case Card::swissGerman:
        {
            switch (suit)
            {
            case 1: return "Acrons";
            case 2: return "Bells";
            case 3: return "Roses";
            case 4: return "Shields";
            }
            break;
        }
        default: // Then its a Latin suit.
        {
            switch (suit)
            {
            case 1: return "Clubs";
            case 2: return "Coins";
            case 3: return "Cups";
            case 4: return "Swords";
            }
        }
    }
}

String Deck::toString ()
{
    String returnString = "Deck: numCards: ";

    returnString += deck.size (); +", ";
    returnString += acesHigh ? "Aces high, " : "Aces low, ";
    returnString += acesHigh ? "Has Jokers\n" : "No Jokers\n";
    
    for (int i = 0; i < deck.size (); ++i)
        returnString += deck[i]->toString () + "\n";
    return returnString;
}

#if DEBUG_MODE

class DeckUnitTest : public UnitTest
{
public:
    //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

    DeckUnitTest () : UnitTest ("Testing KabukiCardGames::Deck class...") {}

    void runTest ()
    {
        logMessage ("Creating testDeck...");

        Deck testDefaultDeck = Deck ();
        
        logMessage (testDefaultDeck.toString ());

        for (int i = 0; i < 7; ++i)
        {
            logMessage ("Creating " + Deck::suitCultureStrings[i] + " Deck...");
            Deck testDeck ((Deck::SuitCulture) i, Deck::acesLow, Deck::doesNotContainJokers);
            logMessage (testDeck.toString ());
        }
        //logMessage (testDeck.toString ());
    }
};
static DeckUnitTest deckUnitTest;
#endif // DEBUG_MODE
